moirae/estimators/online/utils/model.py

```diff
@@ -0,0 +1,153 @@
+
+    @property
+    def num_output_dimensions(self) -> int:
+        return self._num_output_dimensions
+
+
+class JointCellModelInterface(CellModelInterface):
+    """Interface used when the hidden state used by a filter includes the transient states.
+
+    Create the interface by defining
+        - Which portions of the ASOH are used as inputs to function
+        - Values for the ASOH parameters that remain fixed
+        - An example transient state and input to be passed to the function which will be used as a template
+
+    The resultant function will take numpy arrays as inputs and produce numpy arrays as outputs
+
+    Args:
+        model: Model which defines the physics of the system being modeled
+        asoh: Values for all state of health parameters of the model
+        transients: Current values of the transient state of the system
+        input_template: Example input values for the model
+        asoh_inputs: Names of the ASOH parameters to include as part of the hidden state
+    """
+
+    def __init__(self,
+                 model: CellModel,
+                 asoh: HealthVariable,
+                 transients: GeneralContainer,
+                 input_template: InputQuantities,
+                 asoh_inputs: Tuple[str]):
+        super().__init__(model, asoh, transients, input_template)
+
+        # Store the information about the identity of variables in the transient state
+        self.asoh_inputs = asoh_inputs
+        self.num_transients = transients.to_numpy().shape[1]
+        self.num_asoh = asoh.get_parameters(self.asoh_inputs).shape[1]
+
+    @property
+    def num_hidden_dimensions(self) -> int:
+        return self.num_transients + self.num_asoh
+
+    def create_hidden_state(self, asoh: HealthVariable, transients: GeneralContainer) -> np.ndarray:
+        """Transform the state of health and transients states (quantities used by CellModel)
+        into the "hidden state" vector used by the actual filter
+
+        Args:
+            asoh: Values of the ASOH parameter
+            transients: Values of the transient states
+        Returns:
+            A hidden state vector ready for use in a filter
+        """
+
+        return np.concatenate([
+            transients.to_numpy(),
+            asoh.get_parameters(self.asoh_inputs)
+        ], axis=1)
+
+    def create_cell_model_inputs(self, hidden_states: np.ndarray) -> Tuple[HealthVariable, GeneralContainer]:
+        """Convert the hidden states into the forms used by CellModel
+
+        Args:
+            hidden_states: Hidden states as used by the estimator
+        Returns:
+            - ASOH with values from the hidden states
+            - Transients state from the hidden states
+        """
+
+        # Update any parameters for the transient state
+        my_transients = self.transients.model_copy(deep=True)
+        batch_transients = np.repeat(self.transients.to_numpy(), axis=0, repeats=hidden_states.shape[0])

```

Comment:
I'm a bit confused here; why are we repeating the transient state stored in `self`, instead of just doing `my_transient.from_numpy(hidden_states[:, :self.num_transients])`?

Reply:
Good point. I'm betting it's unnecessary and something I did out of abundance of caution over batching